modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp

```diff
@@ -0,0 +1,995 @@
+        .add< TopologyChecker >();
+
+
+TopologyChecker::TopologyChecker()
+    : d_eachStep(initData(&d_eachStep, false, "eachStep", "Check topology at each step"))
+    , l_topology(initLink("topology", "link to the topology container"))
+    , m_topology(nullptr)
+{
+    this->f_listening.setValue(true);
+}
+
+
+TopologyChecker::~TopologyChecker()
+{
+
+}
+
+
+void TopologyChecker::init()
+{
+    if (l_topology.empty())
+    {
+        msg_info() << "link to Topology container should be set to ensure right behavior. First Topology found in current context will be used.";
+        l_topology.set(this->getContext()->getMeshTopologyLink());
+    }
+
+    m_topology = l_topology.get();
+    msg_info() << "Topology path used: '" << l_topology.getLinkedPath() << "'";
+
+    if (m_topology == nullptr)
+    {
+        msg_error() << "No topology component found at path: " << l_topology.getLinkedPath() << ", nor in current context: " << this->getContext()->name;
+        sofa::core::objectmodel::BaseObject::d_componentState.setValue(sofa::core::objectmodel::ComponentState::Invalid);
+        return;
+    }
+}
+
+void TopologyChecker::reinit()
+{
+    checkTopology();
+}
+
+
+bool TopologyChecker::checkTopology()
+{
+    msg_info() << "checkTopology TopologyType: " << parseTopologyElementTypeToString(m_topology->getTopologyType());
+
+    bool result = false;
+    if (m_topology->getTopologyType() == TopologyElementType::HEXAHEDRON)
+        result = checkHexahedronTopology();
+    if (m_topology->getTopologyType() == TopologyElementType::TETRAHEDRON)
+        result = checkTetrahedronTopology();
+    else if (m_topology->getTopologyType() == TopologyElementType::QUAD)
+        result = checkQuadTopology();
+    else if (m_topology->getTopologyType() == TopologyElementType::TRIANGLE)
+        result = checkTriangleTopology();
+    else if (m_topology->getTopologyType() == TopologyElementType::EDGE)
+        result = checkEdgeTopology();
+
+    return result;
+}
+
+
+
+bool TopologyChecker::checkEdgeTopology()
+{
+    // check edge container consistency
+    if (checkEdgeContainer() == false)
+        return false;
+
+    // check cross elements consistency
+    return checkEdgeToVertexCrossContainer();
+}
+
+
+bool TopologyChecker::checkEdgeContainer()
+{
+    sofa::Size nbE = m_topology->getNbEdges();
+    const sofa::core::topology::BaseMeshTopology::SeqEdges& my_edges = m_topology->getEdges();
+
+    if (nbE != my_edges.size())
+    {
+        msg_error() << "CheckEdgeTopology failed: not the good number of edges, getNbEdges returns " << nbE << " whereas edge array size is: " << my_edges.size();
+        return false;
+    }
+
+    // check edge buffer
+    bool res = true;
+    for (sofa::Index i = 0; i < nbE; ++i)
+    {
+        const auto& edge = my_edges[i];
+        if (edge[0] == edge[1]) {
+            msg_error() << "CheckEdgeTopology failed: edge " << i << " has 2 identical vertices: " << edge;
+            res = false;
```

Comment:
Maybe cut the loop once res = false ? (or add the test i < nbE && !res in the for loop)

Reply:
yes, I hesitate, the idea was to catch all the errors... but it quickly become a mess..